### PR TITLE
Allow `cache.set(x, undefined)` to delete the cached item

### LINKED ISSUE #25: Allow `cache.set(x, undefined)` to delete the cached item like `cache.function` does

### PR DESCRIPTION
**Pull request by cheap-glitch**
This is a very basic fix for the issue #25 discovered in sindresorhus/refined-github#3589.

Fixes #25

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -68,8 +68,7 @@ async function get<TValue extends Value>(key: string): Promise<TValue | undefine
 
 async function set<TValue extends Value>(key: string, value: TValue, maxAge: TimeDescriptor = {days: 30}): Promise<TValue> {
 	if (typeof value === 'undefined') {
-		// @ts-ignore This never happens in TS because `value` can't be undefined
-		return;
+		await delete_(key);
```
Comment (fregante) [inline index.ts:71]: I think
```suggestion
		return delete_(key);
```

If not, first you delete it and the. It adds it again later, as undefined.

Comment (cheap-glitch) [reply index.ts:71]: Good point, my bad.

File: index.ts [hunk 2]
```diff
@@ -68,8 +68,7 @@ async function get<TValue extends Value>(key: string): Promise<TValue | undefine
 
 async function set<TValue extends Value>(key: string, value: TValue, maxAge: TimeDescriptor = {days: 30}): Promise<TValue> {
 	if (typeof value === 'undefined') {
-		// @ts-ignore This never happens in TS because `value` can't be undefined
-		return;
+		return delete_(key);
```
Comment (fregante) [inline index.ts:71]: To solve the type issue, you can try:

```suggestion
		await delete_(key);
		return value;
```

**Review by fregante — APPROVED** [review]

**Comment by fregante** [comment]
Thanks for the PR! This is a breaking change and also something that I had initially specifically avoided doing, so I'll probably wait just in case there's a reason not to do it.

**Comment by cheap-glitch** [comment]
> Thanks for the PR! This is a breaking change and also something that I had initially specifically avoided doing, so I'll probably wait just in case there's a reason not to do it.

Maybe because it means `cache.set('myKey')` actually _unsets_ the key instead of setting it, which can arguably be confusing?

**Comment by fregante** [comment]
I think the reasoning was that `undefined` could not be saved, because `undefined` is meant as "missing", even though `.set` is very well capable of storing `undefined` because it wraps it in an object first.

But also, yes, both `.set('a')` and `.set('a', undefined)` deleting `'a'` are kinda weird.

I guess the former more than the latter. The first _could_ throw an error though, at least to avoid using `.set('a')` by mistake instead of `.get('a')`

**Comment by fregante** [comment]
Can you add a `arguments.length` check so that it throws a TypeError when you call `.set('a')`?

**Comment by cheap-glitch** [comment]
> Can you add a `arguments.length` check so that it throws a TypeError when you call `.set('a')`?

Done. Also added a test to cover the new case.

File: index.ts [hunk 3]
```diff
@@ -68,9 +68,13 @@ async function get<TValue extends Value>(key: string): Promise<TValue | undefine
 }
 
 async function set<TValue extends Value>(key: string, value: TValue, maxAge: TimeDescriptor = {days: 30}): Promise<TValue> {
+	if (arguments.length < 2) {
+		throw new TypeError('Need a value');
```
Comment (fregante) [inline index.ts:72]:
```suggestion
		throw new TypeError('Expected a value as the second argument');
```

**Comment by fregante** [comment]
Thanks!

I'm releasing this as a minor after all. The behavior it changes isn't that breaking. Passing `undefined` was a no-op and made no sense in the first place.

File: index.ts [hunk 4]
```diff
@@ -68,18 +68,21 @@ async function get<TValue extends Value>(key: string): Promise<TValue | undefine
 }
 
 async function set<TValue extends Value>(key: string, value: TValue, maxAge: TimeDescriptor = {days: 30}): Promise<TValue> {
```
Comment (fregante) [inline index.ts:70]: I think the only issue here is that `undefined` isn't accepted by the types

```suggestion
async function set<TValue extends Value>(key: string, value: TValue | undefined, maxAge: TimeDescriptor = {days: 30}): Promise<TValue> {
```

I released a prerelease 4.3.0-0 if you want to try it in RGH in the original issue that discussed this

**Pull request merged**